fix issues when cleanupTxn called multiple times

tbg opened the pull request:
now the txnMeta.txnEnd channel is set to nil by whoever gets
there first, everyone after is a no-op. Fixed a small race
in heartbeats regarding that change.

fixes #1823.

MBerhault commented:
LGTM

tamird commented on kv/txn_coord_sender_test.go:
```diff
@@ -746,3 +746,41 @@ func TestTxnDrainingNode(t *testing.T) {
+	txn := newTxn(s.Clock, proto.Key("a"))
+	pReply := &proto.PutResponse{}
+	key := proto.Key("a")
+	call := proto.Call{
```
`proto.PutCall()` might help you here

tamird replied:
oh, transaction. disregard

tamird commented on kv/txn_coord_sender_test.go:
```diff
@@ -746,3 +746,41 @@ func TestTxnDrainingNode(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	if pReply.GoError() != nil {
```
should check `.Error` for symmetry with the below

tschottdorf replied:
done.

tamird commented:
LGTM

bdarnell commented:
LGTM

tamird commented on kv/txn_coord_sender.go:
```diff
@@ -694,7 +695,7 @@ func (tc *TxnCoordSender) cleanupTxn(trace *tracer.Trace, txn proto.Transaction,
 	// Trigger intent resolution and heartbeat shutdown.
 	trace.Event("coordinator stops")
 	txnMeta.txnEnd <- resolved // buffered, so does not block
```
this comment could use a correction since setting to nil makes it block :)

tschottdorf replied:
done.

tbg committed:
fix issues when cleanupTxn called multiple times

now the txnMeta.txnEnd channel is set to nil by whoever gets
there first, everyone after is a no-op. Fixed a small race
in heartbeats regarding that change.

fixes #1823.
